src/lib.rs

```diff
@@ -851,701 +734,171 @@ pub trait Listener<T = ()>: Future<Output = T> + __sealed::Sealed {
-            Self::Waker(waker) => Task::Waker(waker.clone()),
-            #[cfg(all(feature = "std", not(target_family = "wasm")))]
-            Self::Unparker(unparker) => Task::Unparker(unparker.clone()),
-        }
+/// Equivalent to `atomic::fence(Ordering::SeqCst)`, but in some cases faster.
+#[inline]
+fn full_fence() {
+    if cfg!(all(
+        any(target_arch = "x86", target_arch = "x86_64"),
+        not(miri)
+    )) {
+        // HACK(stjepang): On x86 architectures there are two different ways of executing
+        // a `SeqCst` fence.
+        //
+        // 1. `atomic::fence(SeqCst)`, which compiles into a `mfence` instruction.
+        // 2. `_.compare_exchange(_, _, SeqCst, SeqCst)`, which compiles into a `lock cmpxchg` instruction.
+        //
+        // Both instructions have the effect of a full barrier, but empirical benchmarks have shown
+        // that the second one is sometimes a bit faster.
+        //
+        // The ideal solution here would be to use inline assembly, but we're instead creating a
+        // temporary atomic variable and compare-and-exchanging its value. No sane compiler to
+        // x86 platforms is going to optimize this away.
+        atomic::compiler_fence(Ordering::SeqCst);
+        let a = AtomicUsize::new(0);
+        let _ = a.compare_exchange(0, 1, Ordering::SeqCst, Ordering::SeqCst);
+        atomic::compiler_fence(Ordering::SeqCst);
+    } else {
+        atomic::fence(Ordering::SeqCst);
     }
 }
 
-const NEVER_INSERTED_PANIC: &str = "\
-EventListener was not inserted into the linked list, make sure you're not polling \
-EventListener/listener! after it has finished";
-
 #[cfg(not(loom))]
-/// Synchronization primitive implementation.
-mod sync {
+mod loom {
+    pub(crate) use core::cell;
+
     #[cfg(not(feature = "portable-atomic"))]
-    pub(super) use alloc::sync::Arc;
+    pub(crate) use core::sync::atomic;
+
     #[cfg(not(feature = "portable-atomic"))]
-    pub(super) use core::sync::atomic;
+    pub(crate) use alloc::sync::Arc;
 
     #[cfg(feature = "portable-atomic")]
-    pub(super) use portable_atomic_crate as atomic;
-    #[cfg(feature = "portable-atomic")]
-    pub(super) use portable_atomic_util::Arc;
-
-    #[cfg(all(feature = "std", not(loom)))]
-    pub(super) use std::sync::{Mutex, MutexGuard};
+    pub(crate) use portable_atomic_crate as atomic;
 
-    pub(super) trait WithMut {
-        type Output;
+    #[cfg(feature = "portable-atomic")]
+    pub(crate) use portable_atomic_util::Arc;
 
-        fn with_mut<F, R>(&mut self, f: F) -> R
-        where
-            F: FnOnce(&mut Self::Output) -> R;
+    /// Equivalent to `loom::AtomicPtr::with_mut`
+    pub(crate) fn ptr_with_mut<T, R>(
+        ptr: &mut atomic::AtomicPtr<T>,
+        f: impl FnOnce(&mut *mut T) -> R,
+    ) -> R {
+        f(ptr.get_mut())
     }
+}
 
-    impl<T> WithMut for atomic::AtomicPtr<T> {
-        type Output = *mut T;
+#[cfg(loom)]
+mod loom {
+    pub(crate) use loom::cell;
```

Comment:
hmm how does symbol lookup / precedence work here?

Reply:
Only one set of symbols are loaded based on which features are enabled